src/state/observeSpeaker.ts

```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2024 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only
+Please see LICENSE in the repository root for full details.
+*/
+import { Observable, audit, merge, timer, filter, startWith } from "rxjs";
+
+/**
+ * Require 1 second of continuous speaking to become a speaker, and 60 second of
+ * continuous silence to stop being considered a speaker
+ */
+export function observeSpeaker(
+  isSpeakingObservable: Observable<boolean>,
+): Observable<boolean> {
+  return isSpeakingObservable.pipe(
+    audit((s) =>
+      merge(
+        timer(s ? 1000 : 60000),
+        // If the speaking flag resets to its original value during this time,
+        // end the silencing window to stick with that original value
+        isSpeakingObservable.pipe(filter((s1) => s1 !== s)),
+      ),
+    ),
+    startWith(false),
```

Comment:
Thinking through this a bit more seems like we might want to have a `distinctUntilChange` for this one.
Or in general I am wondering how it behaves when we feed multiple speaking=true (or false) events in a row?

I suspect it will emit `undefined` since we dont we will already send the latest audited event with the first timeout but since we starteed another timeout that will then not get any new queued event?

Reply:
I think this suspicion is not true. But the filter is basically a manual `distinctUntilChanged`.
And we can still end up with double emissions. In `observeSpeaker`:
 - we start and **emit false** (or come from true and waited for 60s and emitted false. either option works)
 - set to true (won't emit for 10s because of audit)
 - we get a false so we do emit with `isSpeakingObservable.pipe(filter((s1) => s1 !== s)),` which will **emit false** (because this is the last thing the audit has seen.
 
 So we emit false twice.

Reply:
Do you think the following is easier to understand:
```suggestion
export function observeSpeaker(
  isSpeakingObservable: Observable<boolean>,
): Observable<boolean> {
  const toggleObservable = isSpeakingObservable.pipe(distinctUntilChanged());

  return toggleObservable.pipe(
    audit((s) => merge(timer(s ? 1000 : 60000), toggleObservable)),
    startWith(false),
    distinctUntilChanged(),
```

Reply:
With comments:
```suggestion
export function observeSpeaker(
  isSpeakingObservable: Observable<boolean>,
): Observable<boolean> {
  const toggleObservable = isSpeakingObservable.pipe(distinctUntilChanged());

  return toggleObservable.pipe(
    // Either change to the new value after the timer or re-emit the same value if it toggles back
    // (audit will return the latest (toggled back) value) before the timeout.
    audit((s) => merge(timer(s ? 1000 : 60000), toggleObservable)),
    // Filter the re-emissions (marked as: | ) that happen if we toggle quickly (<1s) from false->true->false|->..
    startWith(false),
    distinctUntilChanged(),
```

Reply:
I originally didn't have the `distinctUntilChanged()` at the end because where the observer is used it is already piped into `this.scope.state()` (see https://github.com/element-hq/element-call/blob/a587956c759d0d58f777828a73f33097a5b1bf1e/src/state/CallViewModel.ts#L251) which does the distinct again.

In https://github.com/element-hq/element-call/commit/a587956c759d0d58f777828a73f33097a5b1bf1e I have made this part of the observable itself.

However, this does mean that we do the `distinctUntilChanged()` twice.

Reply:
I've incorporated your other suggestions too.